onnx/backend/test/case/node/roialign.py

```diff
@@ -161,33 +161,76 @@ def export_roialign():  # type: () -> None
             [
                 [
                     [
-                        [0.4664, 0.4466, 0.3405, 0.5688, 0.6068],
-                        [0.3714, 0.4296, 0.3835, 0.5562, 0.3510],
-                        [0.2768, 0.4883, 0.5222, 0.5528, 0.4171],
-                        [0.4713, 0.4844, 0.6904, 0.4920, 0.8774],
-                        [0.6239, 0.7125, 0.6289, 0.3355, 0.3495],
+                        [0.5178, 0.3434, 0.3229, 0.4474, 0.6344],
```

Comment:
Instead of having just the numbers, is it possible to have code that generates this? That is, some python code that can act as a reference implementation? I guess the original test-case perhaps pre-dates the requirement to use reference implementations to the extent possible.

Reply:
I'm not sure. Maybe [scipy.signal.resample](https://docs.scipy.org/doc/scipy/reference/generated/scipy.signal.resample.html]) or [scipy.ndimage.zoom](https://docs.scipy.org/doc/scipy/reference/generated/scipy.ndimage.zoom.html) would work 🤔 - I only compared against other frameworks like Detectron/PyTorch/TF.

For the 2nd smaller test case at least, I rather like having directly visible numbers, which make the test cases clearer and more immediately comprehensible (saying this as someone who's had to implement multiple ONNX operators for DML, and frequently referred back to Operators.md 😅). For the 1st case with the larger tensors, cutting that sea of numbers and replacing it with a numpy equivalent would be nice.

Reply:
Sure, I agree that for the smaller example, visually matching input/output is useful. But for the larger it is harder. The code could also be specialized for different cases, if it makes the job easier.

Reply:
We only want dependencies on pure numpy for the ONNX backend, right? I couldn't find any function in `numpy` that resamples tensors (and the numpy [interpolation functionality](https://numpy.org/doc/stable/reference/generated/numpy.interp.html) works on a series of *coordinates*, not tensors), and playing with various scipy functions, `scipy.ndimage.zoom` and `scipy.ndimage.affine_transform` were the two that worked as expected, but those are scipy dependencies. :/

```python
import numpy
import scipy
from scipy import ndimage
from scipy import signal
from scipy import misc

numpy.set_printoptions(linewidth=120)

x = numpy.array(
        [[ 0, 1, 2, 3],
         [ 4, 5, 6, 7],
         [ 8, 9,10,11],
         [12,13,14,15]],
        dtype=numpy.float32
    )
print("input:\n", x, sep='')

# ✖numpy.interp - Unsuitable. Takes an array input coordinates, not a tensor.
# https://numpy.org/doc/stable/reference/generated/numpy.interp.html

# ✖scipy.interpolate.interp2d - Unsuitable. Takes an array of input points, not a tensor or 2D image.
# https://docs.scipy.org/doc/scipy/reference/generated/scipy.interpolate.interp2d.html

# ✖scipy.misc.imresize - Unsuitable. Deprecated and removed in later version. Also has dependency on PIL anyway.
# AttributeError: module 'scipy.misc' has no attribute 'imresize'
# https://docs.scipy.org/doc/scipy-1.2.0/reference/generated/scipy.misc.imresize.html
#y = scipy.misc.imresize(x, (8,8), 'bilinear')
#print("\nscipy.misc.imresize\n", y, sep='')

# ✖scipy.signal.resample - Unsuitable. Only takes a single axis.
y = scipy.signal.resample(x, num=8, axis=1)
print("\nscipy.signal.resample:\n", y, sep='')

# ✖scipy.ndimage.zoom grid_mode = False - Unsuitable. Aligns corners rather than pixel centers.
y = scipy.ndimage.zoom(x, zoom=2, order=1, grid_mode=False)
print("\nscipy.ndimage.zoom grid_mode=False:\n", y, sep='')

# ✔scipy.ndimage.zoom grid_mode = True - Suitable. Aligns half pixels.
y = scipy.ndimage.zoom(x, zoom=2, order=1, mode='nearest', grid_mode = True)
print("\nscipy.ndimage.zoom grid_mode=True:\n", y, sep='')

# ✔Suitable.
y = scipy.ndimage.affine_transform(x, matrix=numpy.array([[0.5,0],[0,0.5]]), offset=[-0.25, -0.25], output_shape=(8, 8), order=1, mode="nearest")
print("\nscipy.ndimage.affine_transform:\n", y, sep='')
```

```
input:
[[ 0.  1.  2.  3.]
 [ 4.  5.  6.  7.]
 [ 8.  9. 10. 11.]
 [12. 13. 14. 15.]]

✖scipy.signal.resample:
[[ 0.          0.08578646  1.          1.5         2.          2.9142137   3.          1.5       ]
 [ 4.          4.0857863   5.          5.5         6.          6.9142137   7.          5.5       ]
 [ 8.          8.085787    9.          9.5        10.         10.914213   11.          9.5       ]
 [12.         12.085787   13.         13.5        14.         14.914213   15.         13.5       ]]

✖scipy.ndimage.zoom grid_mode=False:
[[ 0.          0.42857143  0.85714287  1.2857143   1.7142857   2.142857    2.5714285   3.        ]
 [ 1.7142857   2.142857    2.5714285   3.          3.4285715   3.857143    4.285714    4.714286  ]
 [ 3.4285715   3.857143    4.285714    4.714286    5.142857    5.571429    6.          6.428571  ]
 [ 5.142857    5.571429    6.          6.428571    6.857143    7.285714    7.714286    8.142858  ]
 [ 6.857143    7.285714    7.714286    8.142858    8.571428    9.          9.428572    9.857142  ]
 [ 8.571428    9.          9.428572    9.857142   10.285714   10.714286   11.142858   11.571428  ]
 [10.285714   10.714286   11.142858   11.571428   12.         12.428572   12.857142   13.285714  ]
 [12.         12.428572   12.857142   13.285714   13.714286   14.142858   14.571428   15.        ]]

✔scipy.ndimage.zoom grid_mode=True:
[[ 0.    0.25  0.75  1.25  1.75  2.25  2.75  3.  ]
 [ 1.    1.25  1.75  2.25  2.75  3.25  3.75  4.  ]
 [ 3.    3.25  3.75  4.25  4.75  5.25  5.75  6.  ]
 [ 5.    5.25  5.75  6.25  6.75  7.25  7.75  8.  ]
 [ 7.    7.25  7.75  8.25  8.75  9.25  9.75 10.  ]
 [ 9.    9.25  9.75 10.25 10.75 11.25 11.75 12.  ]
 [11.   11.25 11.75 12.25 12.75 13.25 13.75 14.  ]
 [12.   12.25 12.75 13.25 13.75 14.25 14.75 15.  ]]

✔scipy.ndimage.affine_transform:
[[ 0.    0.25  0.75  1.25  1.75  2.25  2.75  3.  ]
 [ 1.    1.25  1.75  2.25  2.75  3.25  3.75  4.  ]
 [ 3.    3.25  3.75  4.25  4.75  5.25  5.75  6.  ]
 [ 5.    5.25  5.75  6.25  6.75  7.25  7.75  8.  ]
 [ 7.    7.25  7.75  8.25  8.75  9.25  9.75 10.  ]
 [ 9.    9.25  9.75 10.25 10.75 11.25 11.75 12.  ]
 [11.   11.25 11.75 12.25 12.75 13.25 13.75 14.  ]
 [12.   12.25 12.75 13.25 13.75 14.25 14.75 15.  ]]
```

Reply:
@askhade : is there any issue or concern with using scipy? Having said that, I don't know if it is worthwhile introducing extra dependencies for this one utility. I took a look at the pooling operations, and they have the same issue (and just have inputs/outputs). Thanks @fdwr for the detailed response. I personally feel it may be okay to leave it as is.

Reply:
usually we either use numpy or custom python functions to generate test data...